src/app.rs

```diff
@@ -476,6 +479,144 @@ impl App {
+        self.event_bus
+            .publish_batch(pending_events)
+            .expect("No one listening on the event bus, committer most likely dead.");
+    }
+
+    #[instrument(level = "info", skip_all)]
+    async fn commit_identity(&self, group_id: usize, commitment: Hash) -> Result<(), ServerError> {
+        // Get a progress lock on the tree for the duration of this operation.
+        let tree = self.merkle_tree.progress().await.map_err(|e| {
+            error!(?e, "Failed to obtain tree lock in commit_identity.");
+            panic!("Sequencer potentially deadlocked, terminating.");
+            #[allow(unreachable_code)]
+            e
+        })?;
+
+        // Fetch next leaf index
+        let identity_index = self.next_leaf.fetch_add(1, Ordering::AcqRel);
+
+        // Send Semaphore transaction
+        let receipt = self
+            .contracts
+            .insert_identity(commitment)
+            .await
+            .map_err(|e| {
+                error!(?e, "Failed to insert identity to contract.");
+                panic!("Failed to submit transaction, state synchronization lost.");
+                #[allow(unreachable_code)]
+                e
+            })?;
+
+        let mut tree = tree.upgrade_to_write().await.map_err(|e| {
+            error!(?e, "Failed to obtain tree lock in insert_identity.");
+            panic!("Sequencer potentially deadlocked, terminating.");
+            #[allow(unreachable_code)]
+            e
+        })?;
+
+        // Update  merkle tree
+        tree.set(identity_index, commitment);
+
+        // Downgrade write lock to progress lock
+        let tree = tree.downgrade_to_progress();
+
+        match receipt.block_number {
+            Some(num) => {
+                info!(
+                    "Identity inserted in block {} at index {}.",
+                    num, identity_index
+                );
+                self.database
+                    .mark_identity_inserted(group_id, &commitment, num.as_usize(), identity_index)
+                    .await?;
+            }
+            None => {
+                panic!("this should not happen?");
+            }
+        }
+
+        // Check tree root
+        if let Err(error) = self.contracts.assert_valid_root(tree.root()).await {
+            error!(
+                computed_root = ?tree.root(),
+                ?error,
+                "Root mismatch between tree and contract."
+            );
+            panic!("Root mismatch between tree and contract.");
+        }
+
+        // Immediately write the tree to storage, before anyone else can
+        // TODO: Store tree in database
+
+        Ok(())
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    fn spawn_identity_committer(self: Arc<Self>) {
+        let mut listener = self.event_bus.subscribe();
+        tokio::spawn(async move {
+            let mut transaction_in_progress = false;
+            let (tx, mut rx) = tokio::sync::mpsc::channel(1);
```

Comment:
These names are not descriptive. I'm also not sure if `mspc` is the appropriate channel type.